Fix drag-n-drop image orientation of preview

krisk opened the pull request:
Fixes #318

This PR addresses the issue of incorrect image preview orientation when drag-n-dropping in the editor.

## Problem

As specified in the issue, the problem is that **exif** image orientation isn't respected.

## Initial investigation

After toying with the initial idea of handling image orientation via CSS transforms, it soon became apparent that there would be several unintended consequences and edge cases to cover. When an image is rotated (via CSS `transform`) its dimensions remain unchanged (i.e, _width_ and _height_ do not automatically flip). This presents problems if, while an image is being uploaded, it is then also edited (such as resized, or captions are added, etc...). These problems are compounded when the transient image is replaced with the persisted copy.

## Solution

Instead, I went the route of re-orienting the actual image, by drawing it into the canvas and applying the appropriate rotation, utilizing the library [JavaScript-Load-Image](https://github.com/blueimp/JavaScript-Load-Image).

At a glance, the following steps are taken:

1. Check image orientation (i.e. read Exif data)
2. If the image orientation needs to be changed, use [JavaScript-Load-Image](https://github.com/blueimp/JavaScript-Load-Image) to draw the re-oriented image.
 - When redrawing, limit the maximum width of the image to be the width of TinyMCE's body. This will result in the constructed image to be of a _reasonable_ file size, otherwise large images will have debilitatingly large data URIs, which _may_ adversely affect UX. Additionally, this helps reduce the probability of stack size limit errors.
3. Retrieve the Data URI, and assign it to the media item URL
4. Build the markup (an image tag) with the Data URI

Additionally, if the image requires re-orientation, then we need to do the following:

1. Block saving from happening as well as the entire auto-save routine. This is because data URIs can quickly exceed length limitations, as well as take up space in `localStorage`.
2. Block the updating of transient images within TinyMCE's editor (i.e, replacing transient images with the persisted ones). This is because when an image is dragged in the drop-zone, if it needs re-orientation then it is **not** actually inserted into the DOM **until** re-orientation finishes.

Once re-orientation completes, we can unblock saving.

To handle the above workflow, I introduced the functions `onBeforeAddMedia` and `onAfterAddMedia`, which are executed as chained promises.

1. Execute `onBeforeAddMedia`: if there's an image which needs re-orientation, block all savings.
2. Then, execute `onAddMedia`: deal with re-orientation, if necessary.
3. Then, execute `onAfterAddMedia`: unblock saving.

## Problems

### Mirrored Images

All orientations are now supported, and the preview correctly shows a properly oriented image. However, I noticed that for all mirrored orientations, while the preview does indeed show the correct re-oriented image, the persisted image is oriented **incorrectly**:

- Orientation 2 (up/mirrored)
- Orientation 4 (down/mirrored)
- Orientation 5 (left/mirrored)
- Orientation 7 (right/mirrored)

I believe this may need to be addressed at the API layer. Otherwise, we'd need to check orientation of the persisted image on the client-side, which may not be the most optimal solution.

### Safari

Safari is still a little finicky with images containing large data-URIs. I'm exploring a fix for this.

## Testing

Exif images are hard to come by, so you can utilize [these](https://github.com/recurser/exif-orientation-examples), and [these](http://www.galloway.me.uk/media/other/EXIF_Orientation_Samples.zip).

I put the entirety of the code behind the feature flag `post-editor/image-reorientation`.

CC: @mdawaffe, @aduth

mdawaffe commented:
Even before the image upload is complete, I see a blob URL for the image in the editor, rather than the data URL we generate with this PR. Is TinyMCE converting it? (Can you track down where that happens?)

Regardless, can we use blob URLs instead of data URLs here? Would that be faster?

mdawaffe commented:
How can we write tests for this behavior?

krisk commented:
@mdawaffe, thanks for the comments :) 

Data URIs are **only** generated for images which need re-orientation. Other images will simply be inserted along with their blob URL. In the HTML editor, you should **not** be seeing blobs for images which lack Exif data, or those which are already properly oriented (if you are, then there's a problem).

But yes, definitely more optimal to use blobs (after re-orientation). Great catch. I'll update the code shortly to reflect this.

TinyMCE does not handle the conversion. Enqueuing of media is handled internally, via execution of [MediaActions.add](https://github.com/krisk/wp-calypso/blob/fix/editor/318-drag-n-drop-orientation/client/lib/media/actions.js#L92), where a `window.File` is [generated](https://github.com/krisk/wp-calypso/blob/fix/editor/318-drag-n-drop-orientation/client/lib/media/actions.js#L131) from the dragged transient image. Said object is then used to [construct](https://github.com/krisk/wp-calypso/blob/fix/editor/318-drag-n-drop-orientation/client/post-editor/media-modal/markup.js#L151) an image tag. This image tag is then inserted into TinyMCE's body by this [plugin](https://github.com/krisk/wp-calypso/blob/fix/editor/318-drag-n-drop-orientation/client/components/tinymce/plugins/media/plugin.jsx#L51).

aduth commented on client/components/tinymce/plugins/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,54 @@
+/**
```
Please keep in mind [coding conventions for JavaScript](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md) in this new file.

```
eslint client/components/tinymce/plugins/media/image-orientation-utils.js 

/Users/andrew/Documents/Code/wp-calypso-krisk/client/components/tinymce/plugins/media/image-orientation-utils.js
  14:3   warning  Expected indentation of 1 tab character but found 0  indent
  14:14  warning  Unexpected space between function name and paren     no-spaced-func
  15:5   warning  Expected indentation of 1 tab character but found 0  indent
  16:7   warning  Expected indentation of 1 tab character but found 0  indent
  21:7   warning  Expected indentation of 1 tab character but found 0  indent
  22:7   warning  Expected indentation of 1 tab character but found 0  indent
  23:7   warning  Expected indentation of 1 tab character but found 0  indent
  26:2   warning  Unnecessary semicolon                                no-extra-semi
  39:3   warning  Expected indentation of 1 tab character but found 0  indent
  40:5   warning  Expected indentation of 1 tab character but found 0  indent
  43:11  warning  Expected indentation of 1 tab character but found 0  indent
  44:13  warning  Expected indentation of 1 tab character but found 0  indent
  46:13  warning  Expected indentation of 1 tab character but found 0  indent
  52:2   warning  Unnecessary semicolon                                no-extra-semi

✖ 14 problems (0 errors, 14 warnings)
```

Depending on your editor, I'd recommend installing both an [EditorConfig](http://editorconfig.org/) and [ESLint](http://eslint.org/docs/user-guide/integrations) plugin to make these issues more obvious, and to not interfere with your personal editor settings outside of Calypso.

aduth commented on client/my-sites/media-library/drop-zone.jsx:
```diff
@@ -35,9 +39,13 @@ module.exports = React.createClass( {
-		MediaActions.add( this.props.site.ID, files );
-		this.props.onAddMedia();
+		this.props.onBeforeAddMedia( files )
+			.then(() => {
```
Will this error when `onBeforeAddMedia` is its default `noop` ?

aduth replied:
Also, needs spaces around function arguments ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls)).

aduth replied:
>Will this error when `onBeforeAddMedia` is its default `noop` ?

Yes it does. The media modal has a separate drop zone implementation, and currently it errors when attempting to drop an image, thus redirecting the browser window to the image itself.

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -91,12 +101,39 @@ export default React.createClass( {
+					if ( orientation > 1 ) {
+						PostActions.blockSave( 'MEDIA_MODAL_IMAGE_TRANSFORMATION' );
+					}
+					return Promise.resolve();
```
Do you need to explicitly return `Promise.resolve()` here since we're already in a promise chain?

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -3,19 +3,29 @@
  */
```
```
eslint client/components/tinymce/plugins/media/drop-zone.jsx 

/Users/andrew/Documents/Code/wp-calypso-krisk/client/components/tinymce/plugins/media/drop-zone.jsx
  147:41  warning  Missing semicolon                                     semi
  162:9   warning  Expected indentation of 7 tab characters but found 8  indent
  163:8   warning  Expected indentation of 6 tab characters but found 7  indent
  171:2   warning  Mixed spaces and tabs                                 no-mixed-spaces-and-tabs
  172:2   warning  Mixed spaces and tabs                                 no-mixed-spaces-and-tabs
  173:2   warning  Mixed spaces and tabs                                 no-mixed-spaces-and-tabs
  175:11  warning  Missing semicolon                                     semi
  178:6   warning  Missing semicolon                                     semi
  184:72  warning  Missing semicolon                                     semi
  188:55  warning  Missing semicolon                                     semi
  189:28  warning  Missing semicolon                                     semi
  194:27  warning  Missing semicolon                                     semi

✖ 12 problems (0 errors, 12 warnings)
```

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -91,12 +101,39 @@ export default React.createClass( {
+		// finishes.
+		// Once re-orientation completes, we can unblock saving (handled in `afterInsertMedia`)
+		if ( isSingleImage ) {
+			return getImageOrientation( files[0] )
```
Technically there's an exception for spacing of literal arguments to arrays ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#arrays-and-function-calls)) which is occasionally ignored, but we should at least be consistent within the same file (see a few lines above where you'd added spaces to `files[ 0 ]`)

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -91,12 +101,39 @@ export default React.createClass( {
+		// re-orientation then it is NOT inserted into the DOM until re-orientation
+		// finishes.
+		// Once re-orientation completes, we can unblock saving (handled in `afterInsertMedia`)
+		if ( isSingleImage ) {
```
To keep indentation lower, usually we recommend taking the approach of returning early. In this case, you could do:

```js
if ( ! isSingleImage ) {
	return Promise.resolve();
}
```

Not only does this keep the indentation of the rest of the function one level lower, but also makes the `else` case much more obvious.

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
 			// For single image upload, insert into post content, blocking save
 			// until the image has finished upload
-			if ( selectedItems[ 0 ].transient ) {
+			let selectedItem = selectedItems[ 0 ]
```
If you're not reassigning `selectedItem`, should be declared with `const` ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#variable-declarations)).

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
+				let handleImageOrientation = ( orientation ) => {
+					// If the orientation is `1` (i.e., "top"), then no further orientation
+					// changes are necessary.
+					if ( orientation > 1 ) {
```
Similar to previous note about early return, I think you could do the same here:

```js
if ( 1 === orientation ) {
	return Promise.resolve( selectedItem );
}
```

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
-			onRenderModal( { visible: true } );
+			analytics.mc.bumpStat( 'editor_upload_via', 'editor_drop' );
+
+			if ( config.isEnabled( 'post-editor/image-reorientation' ) ) {
```
Early return:

```js
if ( ! config.isEnabled( 'post-editor/image-reorientation' ) ) {
	onInsertMedia( markup.get( selectedItem ) );
	MediaActions.setLibrarySelectedItems( site.ID, [] )
	return Promise.resolve();
}
```

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
 		if ( isSingleImage && ! MediaValidationStore.hasErrors( site.ID ) ) {
```
Early return:

```js
if ( ! isSingleImage || MediaValidationStore.hasErrors( site.ID ) ) {
	// Show the media modal list view if not single image or if errors
	onRenderModal( { visible: true } );
	return Promise.resolve();
}
```

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -107,18 +144,67 @@ export default React.createClass( {
+								maxWidth: editor.getBody().clientWidth
+							} )
+							.then( url => {
+								selectedItem.URL = url;
```
Still working my way through the logic here, but at a glance, this mutation of the media state would seem to violate the unidirectional data flow of Flux: https://facebook.github.io/flux/docs/overview.html#structure-and-data-flow

Could we return a new object without mutating the original?

```js
return Object.assign( {}, selectedItem, { URL: url } );
```

aduth commented on client/components/tinymce/plugins/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,54 @@
+export function getImageOrientation( file ) {
+  return new Promise ( resolve => {
+    loadImage.parseMetaData( file, data => {
+      resolve( data.exif ? data.exif.get( 'Orientation' ) : -1 );
```
Is it not sufficient to return a fallback of the default orientation `1`, or do we need to specially account for cases where it couldn't be determined?

krisk replied:
I opted to use `-1` to indicate the absence of exif data, I could change it to simply default to `1`, but I wanted to be more explicit.

aduth commented on client/components/tinymce/plugins/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,54 @@
+    loadImage.parseMetaData( file, data => {
+      resolve( data.exif ? data.exif.get( 'Orientation' ) : -1 );
+    }, {
+      // Save on computation by disabling the following Exif properties
```
Good thinking to skip these 👍

aduth commented on client/components/tinymce/plugins/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,54 @@
+  return new Promise( ( resolve, reject ) => {
+    loadImage(
+        url,
+        res => {
```
We tend to avoid abbreviations in favor of their more verbose but legible equivalents, in this case `result` (case in point, I almost recommended replacing with `response` 😆 ).

krisk replied:
Fair point 😄  Will change it. 

I'm toying with the idea of chaning `url` to `UniformResourceLocator` as well 😆, j/k)

aduth replied:
>I'm toying with the idea of changing `url` to `UniformResourceLocator` as well 😆, j/k)

Lol!

aduth commented on client/components/tinymce/plugins/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,54 @@
+            });
+          }
+        },
+        assign( { orientation: orientation }, options )
```
Tip: ES6 includes a shorthand convenience for constructing objects this way: 

```js
assign( { orientation }, options )
```

([reference](https://github.com/lukehoban/es6features#enhanced-object-literals))

aduth commented on client/components/tinymce/plugins/media/drop-zone.jsx:
```diff
@@ -91,12 +101,39 @@ export default React.createClass( {
+			return getImageOrientation( files[0] )
+				.then( orientation => {
+					if ( orientation > 1 ) {
+						PostActions.blockSave( 'MEDIA_MODAL_IMAGE_TRANSFORMATION' );
```
If the reorientation only happens while an upload is in progress (when saving is already blocked), do we need to block it again?

aduth commented:
Sorry if this throws a wrench in the plan here, but the reorientation issue is not exclusive to drag-and-drop. We also allow users to upload an image from the media modal and insert that image before it's finished uploading. The following steps insert an image that is oriented incorrectly while on this branch:

1. Navigate to the post editor
2. Click the media button in the editor toolbar
3. Click Add New button
4. Select image with image orientation to upload
5. While still uploading, click Insert in modal action bar *
6. Note that image is oriented incorrectly

_*_ Tip: If you're using Chrome, you can slow down your network speed using throttling from the Network tab to give yourself sufficient time to click the insert button:

![throttle](https://cloud.githubusercontent.com/assets/1779930/14856365/bf129878-0c65-11e6-91f1-3a4ab96b7c8f.png)

krisk commented:
Thanks @aduth 😄  for the feedback! I'll be addressing these soon.

mdawaffe commented:
@aduth do you have any concerns about the size of the new libraries?

aduth commented:
>@aduth do you have any concerns about the size of the new libraries?

Usually I'm sensitive to package sizes, especially as it affects the editor (which is already a very large bundle), but in this case the libraries seem reasonably small. That said, I'm sure an argument could be made that this issue is something of an edge case, so any addition would require greater scrutiny.

Approximations:
- [load-image.all.min.js](https://github.com/blueimp/JavaScript-Load-Image/blob/master/js/load-image.all.min.js) (15.8kb before gzip)
- [canvas-to-blob.min.js](https://github.com/blueimp/JavaScript-Canvas-to-Blob/blob/master/js/canvas-to-blob.min.js) (1.2kb before gzip)

Could check to see actual effect on bundle size, but this should be relatively accurate since the libraries have no dependencies of their own.

I've wanted to explore options for splitting off media as a separate bundle from the editor, but they're fairly tightly connected (especially with drag-and-drop being present on the entire screen), so it's not a straight-forward task.

krisk commented:
@aduth, I'm thinking of simplifying the entire re-orientation flow by limiting the changes to a single place, in [`MediaActions.add`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/media/actions.js#L93). Within this function we can do the necessary image manipulation logic (if applicable).

Advantages of this approach:

- We can create the new blob URL before the media item is created, added to the media store, and events are dispatched. This is quite advantageous, since now what is in the media store would indeed be the final copy.
- All cases (so far) would be covered: drag and drop, inserts via media modal (what you pointed out), as well as thumbnail preview in the media-modal (another case which was broken).

I do have a working copy of this solution, but I wanted to get some critique of the approach before I submit the PR 😄

aduth commented:
@krisk From what I recall, what you describe is similar to the path I started to explore. Given the nature of Flux one-way data flow, we could also implement this in such a way that we simply dispatch an updated, correctly-oriented image to replace the transient (while still uploading). I agree it may be a bit awkward for this logic to live in `MediaActions`, but that seems most obvious to me, unless we implemented a separate library that would listen for the store changes and reorient newly received transient images.

Flow I have in mind:
- Image upload initiated (dispatch transient image received)
- Async image reorientation (dispatch reoriented transient image after complete, replaces transient)
- Image upload finishes (dispatch uploaded image to replace transient image)

I think this differs slightly from what you suggested, since it sounds in your description as though the transient image would not be dispatched to the store until after reorientation completes.

krisk commented:
@aduth, thanks for the input 😃 

I had initially implemented it using the dual dispatch approach, however, there was one key issue: 

The moment an image is added via `MediaActions.add`, the event is immediately dispatched, and the image is added to the store. Said image is immediately used for rendering the preview. But of course, the image may be an incorrectly oriented one (because re-orientation happens simultaneously, which eventually dispatches another event). Ergo, we'd have to essentially skip its rendering until the correctly re-oriented one is added to the store instead.  

I didn't find the implementation palatable as it requires checking for some "usable" state (in [plugin.jsx](https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/media/plugin.jsx#L134)) before the image is rendered.

Instead, I modified `MediaActions.add` into two main flows:

#### Part A: create transient media

1. Image is added via  `MediaActions.add`
2. Fix image orientation, if applicable.
3. Dispatch transient image received

Indeed, the event is not dispatched until the image re-orientation is complete.

#### Part B: upload media

At this point, all transient media is in usable state. That is, all images which needed re-orientation are correctly re-oriented.

1. For each media, upload it (ordered, series upload)
2. Dispatch event to replace transient with persisted

The main benefits of this approach:

- Image logic contained in one place
- No need to check for usable state before rendering the preview
- Media store contains the actual media that is and would be rendered

I updated the PR with these changes 😄

aduth commented on client/lib/media/test/actions.js:
```diff
@@ -184,91 +192,113 @@ describe( 'MediaActions', function() {
 					parent_id: 200
 				} );
+
+				done();
```
Did you find you needed to call `done` in these tests? Mocha is happy to work with promise return values in lieu of an explicit callback.

aduth commented on client/my-sites/media-library/upload-url.jsx:
```diff
@@ -49,11 +49,13 @@ export default React.createClass( {
 		MediaActions.clearValidationErrors( this.props.site.ID );
-		MediaActions.add( this.props.site.ID, this.state.value );
+		MediaActions.add( this.props.site.ID, this.state.value )
+			.then( () => {
```
Do we need to react to the promise chain of the action? This really breaks the one-way data flow of Flux. We should be reacting to the change events emitted by the store itself, not the progress of a return value from the action creator. Were there complications in doing so?

krisk replied:
The problem is that `onAddMedia` (the following line) immediately interrogates the media store for the image that needs to be inserted as preview. If the image needs to be re-oriented, it needs to wait for re-orientation to complete.

aduth commented on client/lib/media/actions.js:
```diff
@@ -99,91 +100,139 @@ MediaActions.add = function( siteId, files ) {
+
+	const create = createMedia();
+
+	if ( options.waitForUpload ) {
```
Are we or do we plan to use `waitForUpload` ?

krisk replied:
I wanted to be consistent with what existed before. Previously, the final promise resolved when the uploading was complete. The tests use `waitForUpload`.

aduth commented on client/lib/media/actions.js:
```diff
@@ -99,91 +100,139 @@ MediaActions.add = function( siteId, files ) {
 		} );
-	}, Promise.resolve() );
+
+		return Promise.all( mediaPromises ).then( () => media );
```
Rather than building an array via push, we can leverage the fact that `Promise.all` calls its thenable with an array of results from each promise, which is exactly how you've constructed this anyways.

e.g.
```js
Promise.all( [
	Promise.resolve( { ID: 1 } ),
	Promise.resolve( { ID: 2 } ),
	Promise.resolve( { ID: 3 } )
] ).then( ( media ) => { /* ... */ } );
```

So if you were to change `media.push( {} )` to `return {};`, you could drop the `.then()` here and eliminate the `media` variable.

aduth commented:
As-is, are we waiting for the image to be reoriented before starting the upload? Should we be concerned about that?

Can you think of how we might test the `image-orientation-utils.js` file?

>The moment an image is added via `MediaActions.add`, the event is dispatched, the image is added to the store, and is **immediately** used for rendering the preview. 

I'd probably agree with this being an issue, mostly from a UX perspective. It'd be pretty jarring to see the incorrectly rotated image very briefly, only to be replaced after the client-side reorienting completes.

aduth commented on client/lib/media/image-orientation-utils.js:
```diff
@@ -0,0 +1,76 @@
+			// If the orientation is <= 1 (where 1 is "top", and anything less
+			// than that means there was no Exif data), then no further orientation
+			// changes are necessary.
+			if ( 1 >= orientation ) {
```
Not sure if you were going for "Yoda" style here, but (a) this tends not to apply to greater/less than operators because they're difficult to read ([reference](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#yoda-conditions)) and (b) is optional for Calypso anyways ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#yoda-conditions)). I tend to use them myself, mostly for consistency when moving between WordPress PHP and JavaScript.

aduth commented:
Did you intend to close this PR?
